### PR TITLE
Add fused scale mask bias softmax

### PR DESCRIPTION
**Pull request by ofhwei**
该pr主要是对 `softmax(x * scale + mask + bias)` 多步操作进行融合，减少内存访问次数以提升效率，适用于通用的attention场景: 其中x为query和key矩阵乘的结果，形状一般为[batch_size, num_heads, seq_len_q, seq_len_kv]; scale=sqrt(head_size); mask和bias分别为[batch_size, 1, 1, seq_len_k] 和 [1, num_heads, seq_len_q, seq_len_kv]。
对于某些场景，输入的mask的形状可能也不同([如alphafold](https://github.com/Oneflow-Inc/OneTeam/issues/1717#issuecomment-1290026345))：
1. global attn中x.shape=[b, h, s], mask.shape=[b, 1, s], 
2. template_pointwise_attn中x.shape=[s, s, 1, n_templ], mask.shape=[1, 1, 1, n_templ]等。

针对上述特殊情形，该pr也做了一些针对性处理。

对于半精度类型(fp16和bf16)可以使用flash attention替换，但目前flash attention 还不支持tf32和fp32，后续应该还会逐渐完善。

### CODE REVIEW

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+Maybe<void> FusedScaleMaskBiasSoftmax::Capture(FusedScaleMaskBiasSoftmaxCaptureState* ctx,
+                                               const TensorTuple& inputs,
+                                               const TensorTuple& outputs,
+                                               const AttrMap& attrs) const {
+  ComposedAttrMap composed_attrs(attrs, base_attrs_);
+  ctx->input_requires_grad = inputs.at(0)->requires_grad();
+  if (!ctx->input_requires_grad) { return Maybe<void>::Ok(); }
+
+  if (inputs.size() == 3) ctx->bias_requires_grad = inputs.at(2)->requires_grad();
```
Comment (Ldpe2G) [inline oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp:63]: 这一句应该放到 61 行之前？

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                                             TensorTuple* in_grads) const {
+  CHECK_EQ_OR_RETURN(out_grads.size(), 1);  // dy
+  if (!ctx->input_requires_grad && !ctx->bias_requires_grad) { return Maybe<void>::Ok(); }
+  in_grads->resize(ctx->input_size);
+
+  const std::shared_ptr<oneflow::one::Tensor>& y = ctx->SavedTensors().at(0);
+  const std::shared_ptr<oneflow::one::Tensor>& input_grad =
+      JUST(functional::FusedScaleMaskBiasSoftmaxGrad(y, out_grads.at(0), ctx->scale));
+
+  in_grads->at(0) = input_grad;
```
Comment (Ldpe2G) [inline oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp:80]: 76~80 应该也要放在 `if (ctx->input_requires_grad)` 条件判断里

**Comment by Ldpe2G** [comment]
建议这里加一下文档，描述一下 ` fused_msa_softmax` 和一般的 softmax 的区别是啥

File: oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+/*static*/ auto FusedScaleMaskBiasSoftmaxOp::InferLogicalTensorDesc(user_op::InferContext* ctx)
+    -> Maybe<void> {
+  const float scale = ctx->Attr<float>("scale");
+  CHECK_LE_OR_RETURN(scale, 1.);
+
+  const Shape& x_shape = ctx->InputShape("x", 0);
+  const Shape& mask_shape = ctx->InputShape("mask", 0);
+  CHECK_OR_RETURN(x_shape[-1] == mask_shape[-1]
```
Comment (Ldpe2G) [inline oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp:46]: 对于取 shape 元素，建议统一取法，要不都用 `At` 要不就都用  `[]`。
然后这里可以解释一下 shape 的判断逻辑是什么，对这个算子的功能之前不了解

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp (hunk 1 above)
Comment (ofhwei) [reply oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp:61]: 好的，已经修改。

File: oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp (hunk 3 above)
Comment (ofhwei) [reply oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp:46]: 好的，统一改成 `[]` 了。在[前面](https://github.com/Oneflow-Inc/oneflow/pull/9867#issue-1584002873)也已增加了一些说明，现在fused_msa_softmax改成FusedScaleMaskBiasSoftmax了，更加通用一些。

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                                             TensorTuple* in_grads) const {
+  CHECK_EQ_OR_RETURN(out_grads.size(), 1);  // dy
+  if (!ctx->input_requires_grad && !ctx->bias_requires_grad) { return Maybe<void>::Ok(); }
+  in_grads->resize(ctx->input_size);
+
+  const std::shared_ptr<oneflow::one::Tensor>& y = ctx->SavedTensors().at(0);
+  const std::shared_ptr<oneflow::one::Tensor>& input_grad =
+      JUST(functional::FusedScaleMaskBiasSoftmaxGrad(y, out_grads.at(0), ctx->scale));
+
+  if (ctx->input_requires_grad) in_grads->at(0) = input_grad;
```
Comment (Ldpe2G) [inline oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp:80]: 应该 76~77 都放到这个 `if` 判断里，如果 `ctx->input_requires_grad == false` ， 那就根本不需要调用 input_grad 的计算，而且上面65行也不用 saveTensorForBackward 了。

File: oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp (hunk 3 above)
Comment (Ldpe2G) [reply oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp:46]: 好的

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp (hunk 4 above)
Comment (ofhwei) [reply oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp:80]: 假如73行的条件`if (!ctx->input_requires_grad && !ctx->bias_requires_grad)` 不成立，说明`input`或者`bias`至少有一个需要求导，这样76-77是不可少的。在85行求bias_grad也有用到这个 `input_grad`

Comment (Ldpe2G) [reply oneflow/core/autograd/gradient_funcs/fused_scale_mask_bias_softmax.cpp:80]: 好的，了解了

File: oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  }
+  ctx->SetOutputDType("out", 0, query_type);
+  return Maybe<void>::Ok();
+}
+
+/*static*/ auto FusedScaleMaskBiasSoftmaxOp::InferLogicalTensorDesc(user_op::InferContext* ctx)
+    -> Maybe<void> {
+  const float scale = ctx->Attr<float>("scale");
+  CHECK_LE_OR_RETURN(scale, 1.);
+
```
Comment (Ldpe2G) [inline oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp:43]: 对于下面的形状检查的逻辑，可以加些注释，简单讲解一下 x, mask 和 bias 支持哪些形状

**Review by Ldpe2G — APPROVED** [review]

File: oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp (hunk 5 above)
Comment (ofhwei) [reply oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp:43]: 好的👌

File: oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  const Shape& x_shape = ctx->InputShape("x", 0);
+  const Shape& mask_shape = ctx->InputShape("mask", 0);
+  CHECK_OR_RETURN(x_shape[-1] == mask_shape[-1]
+                  && (x_shape[0] == mask_shape[0] || mask_shape[0] == 1));
```
Comment (yz-chen18) [inline oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp:47]: mask的第一维（batch_size或ensemble_batch）似乎应该始终与x的第一维相等，不存在mask_shape第一维等于1需要广播的情况

Comment (ofhwei) [reply oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp]: 是的，已修改

File: oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    const Shape& bias_shape = ctx->InputShape("bias", 0);
+    CHECK_OR_RETURN(mask_shape[-1] == bias_shape[-1]);
+    CHECK_OR_RETURN(mask_shape[0] == bias_shape[0] || bias_shape[0] == 1);
+    for (int i = 1; i < x_shape.NumAxes() - 1; i++) {
+      CHECK_OR_RETURN((mask_shape[i] == 1 || bias_shape[i] == 1)
+                      && mask_shape[i] * bias_shape[i] == x_shape[i]);
+    }
+  } else {
+    auto axes = x_shape.NumAxes();
+    CHECK_OR_RETURN(mask_shape[-1] == x_shape[-1]);
```
[the remark this answers is not in the document]
Comment (ofhwei) [reply oneflow/user/ops/fused_scale_mask_bias_softmax_op.cpp]: 已删

**Review by yz-chen18 — APPROVED** [review]
